src/classes/RD2_RecurringDonations_TDTM.cls

```diff
@@ -126,6 +127,37 @@ public class RD2_RecurringDonations_TDTM extends TDTM_Runnable {
         }
     }
 
+    /***
+    * @description Data Transformation Actions for Before Insert and Update Trigger Actions:
+    * - If RecurringType is Open and NumOfPlannedInstallments = 1, set NumOfPlannedInstallments to null
+    * - If RecurringType is Fixed and NumberOfPaidInstallments is >= NumOfPlannedInstallments Close RD
+    * @param rds Trigger.new records
+    * @param oldRds Trigger.old records
+    * @param triggerAction BeforeInsert, BeforeUpdate, etc.
+    */
+    public void handleDataTransformationInBeforeContext(List<npe03__Recurring_Donation__c> rds,
+            List<npe03__Recurring_Donation__c> oldRds, TDTM_Runnable.Action triggerAction
+    ) {
+
+        if (triggerAction != TDTM_Runnable.Action.BeforeInsert &&
+                triggerAction != TDTM_Runnable.Action.BeforeUpdate) {
+            return;
+        }
+
+        for (npe03__Recurring_Donation__c rd : rds) {
```

Comment:
Great questions. My initial interpretation is that validation should happen first. You raise a really good point about `handleRDSchedules()`. That method actually does more than just schedule work already. It was the name that threw me off. I didn't even look into the code because I didn't want to mix up these _before_ actions with schedule stuff. So, for this I think it would make sense to combine the methods as you've indicated **and** to rename the method to `handleRDScheduleAndDataCalculations`. Not a perfect name, but it does indicate that there's more happening than just schedule work.

Reply:
What do you think about grouping all before context into one method and leave handling of schedules for the after trigger context?

For example the `run()` method will contain now this segment
```
autogenerateNames(rds, oldRds, triggerAction);
validate(rds, oldRds, triggerAction);
transformDataInBeforeContext(rds, oldRds, triggerAction);
handleSchedulesInAfterContext(rds, oldRds, triggerAction);
```

where 
```
private void transformDataInBeforeContext(List<npe03__Recurring_Donation__c> rds,
        List<npe03__Recurring_Donation__c> oldRds, TDTM_Runnable.Action triggerAction
    ) {
        if (triggerAction != TDTM_Runnable.Action.BeforeInsert	
            && triggerAction != TDTM_Runnable.Action.BeforeUpdate	
        ) {	
            return;	
        }

        Map<Id, List<RecurringDonationSchedule__c>> schedulesByRdId = triggerAction == TDTM_Runnable.Action.BeforeUpdate
            ? scheduleService.getAllRecurringDonationSchedules(rds, oldRds)
            : new Map<Id, List<RecurringDonationSchedule__c>>();

        for (npe03__Recurring_Donation__c rd : rds) {
            RD2_OpportunityEvaluationService.RecurringDonation rdRecord = new RD2_OpportunityEvaluationService.RecurringDonation(rd);

            if (rdRecord.isOpenEnded() && rd.npe03__Installments__c == 1) {
                rd.npe03__Installments__c = null;
            } 

            if (triggerAction == TDTM_Runnable.Action.BeforeInsert) {
                rd.npe03__Next_Payment_Date__c = scheduleService.getNextDonationDate(rd);

            } else if (triggerAction == TDTM_Runnable.Action.BeforeUpdate) {
                rdRecord.reviseStatusForFixedLength();

                rd.npe03__Next_Payment_Date__c = scheduleService.getNextDonationDate(schedulesByRdId.get(rd.Id));             
            }
        }
    }
```

and `handleRDScheduleAndDataCalculations()` becomes:
```
private void handleSchedulesInAfterContext(List<SObject> rds, List<SObject> oldRds, TDTM_Runnable.Action triggerAction) {

        if (triggerAction == TDTM_Runnable.Action.AfterInsert) {
            buildNewSchedules(rds);

        } else if (triggerAction == TDTM_Runnable.Action.AfterUpdate) {
            synchronizeSchedulesAfterRDUpdate(rds, oldRds);
        }
    }
```

The `calculateNextDonationDates() `and `calculateNextDonationDateBeforeUpdate()` can be removed.

Reply:
I see what you're getting at here. There's some duplicate logic and loops for both similar and somewhat different purposes, but both are executing in the _before_ context. Combining makes good sense. Thanks for drilling into the full detail of the class logic to identify this.